src/binary-reader-linker.c

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2016 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "binary-reader-linker.h"
+
+#include "binary-reader.h"
+#include "wasm-link.h"
+
+#define RELOC_SIZE 5
+
+typedef struct Context {
+  WasmAllocator* allocator;
+  InputBinary* binary;
+
+  WasmStringSlice import_name;
+  Section* current_section;
+} Context;
+
+static WasmResult on_reloc(WasmRelocType type,
+                           uint32_t section_index,
+                           uint32_t offset,
+                           void* user_data) {
+  Context* ctx = user_data;
+  InputBinary* binary = ctx->binary;
+  if (section_index >= binary->sections.size) {
+    WASM_FATAL("invalid section index: %d\n", section_index);
+  }
+  Section* sec = &binary->sections.data[section_index];
```

Comment:
I changed it from a one-based to zero-based index.   I got feedback from dan that maybe we should use the section_code instead though so this might change again.